docs/error-handling.md

```diff
@@ -0,0 +1,190 @@
+# Error handling and logging
+
+## Logging errors
+
+The principle is that errors are logged when they are handled. If you
+just propagate an error to the caller in a function, you don't need to
+log it; the caller will. But if you "swallow" an error in a function,
+you *must* log it (if it needs to be logged at all).
+
+For example:
+
+```rust
+fn read_motd_file() -> std::io::Result<String> {
+    let mut f = File::open("/etc/motd")?;
+    let mut result = String::new();
+    f.read_to_string(&mut result)?;
+    result
+}
+```
+
+Opening or reading the file could fail, but there is no need to log
+the error here. The function merely propagates the error to the
+caller, and it is up to the caller to log the error or propagate it
+further, if the failure is not expected. But if, for example, it is
+normal that the "/etc/motd" file doesn't exist, the caller can choose
+to silently ignore the error, or log it in INFO or DEBUG level:
+
+```rust
+fn get_message_of_the_day() -> String {
+    // Get the motd from /etc/motd, or return the default proverb
+    match read_motd_file() {
+        Ok(motd) => motd,
+        Err(err)  => {
+            // It's normal that /etc/motd doesn't exist, but if we fail to
+            // read it for some other reason, that's unexpected. The message
+            // of the day isn't very important though, so we just WARN and
+            // continue with the default in any case.
+            if err.kind() != std::io::ErrorKind::NotFound {
+                 tracing::warn!("could not read \"/etc/motd\": {err:?}");
```

Comment:
Instead of just formatting the message, it is helpful for generating structured logs when relevant bits of information are recorded to separate fields. This particular example might not be the prime one for that, but here we could do:
```rust
warn!(file = "/etc/motd", "could not read {file}: {err:?}");
```

So the log output becomes(plain format):
```console
2022-10-23T15:09:32.007339Z  WARN could not read /etc/motd: PermissionDenied file=/etc/motd
```

and with json formatting we would have:
```json
{"timestamp":"2022-10-23T15:09:32.007339Z","level":"WARN","fields":{"message":"could not read /etc/motd: PermissionDenied","file":"/etc/motd"}}
```

We could distill a couple sentences about structuring and then have our examples follow that.

Reply:
Re: my other https://github.com/neondatabase/neon/pull/2681#issuecomment-1288554825, I'd go as far as not including any variable context in the message:

```rust
warn!(?file, ?err, "read failed");
```

But will need to do some experimentation on what this would look like.

Reply:
I would like to agree and partially do, but wouldn't suggest completely against having some context in the message either, because it's a bit less tool agnostic in sense `message` field becomes a bit less useful and then there needs to be a view for each type of logs. So would write the message for humans and separate the fields for computers, so that they can make things more accessible for humans.